File: dashboards/go/handlers/skills.go
package handlers

import (
	"agenticmail-dashboard/services"
	"agenticmail-dashboard/templates"
	"fmt"
	"net/http"
)

// HandleSkills handles the skills page (GET), and skill enable/disable/uninstall (POST).
func HandleSkills(w http.ResponseWriter, r *http.Request) {
	s := services.GetSession(r)

	if r.Method == "POST" {
		r.ParseForm()
		action := r.FormValue("action")
		skillID := r.FormValue("id")
		body := map[string]string{"orgId": "default"}
		switch action {
		case "enable":
			services.APICall("/api/engine/community/skills/"+skillID+"/enable", "PUT", s.Token, body)
		case "disable":
			services.APICall("/api/engine/community/skills/"+skillID+"/disable", "PUT", s.Token, body)
		case "uninstall":
			services.APICall("/api/engine/community/skills/"+skillID+"/uninstall", "DELETE", s.Token, body)
		}
		http.Redirect(w, r, "/skills", http.StatusFound)
		return
	}

	builtinData, _ := services.APICall("/api/engine/skills/by-category", "GET", s.Token, nil)
	installedData, _ := services.APICall("/api/engine/community/installed?orgId=default", "GET", s.Token, nil)

	// Builtin skills grid
	var builtinHTML string
	if categories, ok := builtinData["categories"].(map[string]interface{}); ok && len(categories) > 0 {
		cards := ""
		for catName, catSkills := range categories {
			if skills, ok := catSkills.([]interface{}); ok {
				for _, sk := range skills {
					skill := sk.(map[string]interface{})
					name := templates.StrVal(skill, "name")
					desc := templates.StrVal(skill, "description")
					if desc == "" {
						desc = "No description"
					}
					cards += fmt.Sprintf(`<div style="background:var(--bg);border:1px solid var(--border);border-radius:8px;padding:16px">
<div style="display:flex;justify-content:space-between;align-items:start;margin-bottom:8px">
<strong style="font-size:13px">%s</strong>%s</div>
<div style="font-size:12px;color:var(--dim);line-height:1.5">%s</div></div>`,
						templates.Esc(name), templates.Badge(catName), templates.Esc(desc))
				}
			}
		}
		builtinHTML = `<div style="display:grid;grid-template-columns:repeat(auto-fill,minmax(280px,1fr));gap:12px">` + cards + `</div>`
	} else if skills, ok := builtinData["skills"].([]interface{}); ok && len(skills) > 0 {
		cards := ""
		for _, sk := range skills {
			skill := sk.(map[string]interface{})
			name := templates.StrVal(skill, "name")
			desc := templates.StrVal(skill, "description")
			category := templates.StrVal(skill, "category")
			if desc == "" {
				desc = "No description"
			}
			if category == "" {
				category = "general"
			}
			cards += fmt.Sprintf(`<div style="background:var(--bg);border:1px solid var(--border);border-radius:8px;padding:16px">
<div style="display:flex;justify-content:space-between;align-items:start;margin-bottom:8px">
<strong style="font-size:13px">%s</strong>%s</div>
<div style="font-size:12px;color:var(--dim);line-height:1.5">%s</div></div>`,
				templates.Esc(name), templates.Badge(category), templates.Esc(desc))
		}
		builtinHTML = `<div style="display:grid;grid-template-columns:repeat(auto-fill,minmax(280px,1fr));gap:12px">` + cards + `</div>`
	} else {
		builtinHTML = `<div class="empty"><div class="empty-i">⚡</div>No builtin skills available</div>`
	}

	// Installed community skills table
	var installedHTML string
	var installedList []interface{}
	if skills, ok := installedData["skills"].([]interface{}); ok {
		installedList = skills
	} else if skills, ok := installedData["installed"].([]interface{}); ok {
		installedList = skills
	}

	if len(installedList) > 0 {
		rows := ""
		for _, sk := range installedList {
			skill := sk.(map[string]interface{})
			name := templates.StrVal(skill, "name")
			desc := templates.StrVal(skill, "description")
			status := templates.StrVal(skill, "status")
			if status == "" {
				status = "enabled"
			}
			id := templates.StrVal(skill, "id")

			toggleAction := "disable"
			toggleLabel := "Disable"
			if status == "disabled" {
				toggleAction = "enable"
				toggleLabel = "Enable"
			}

			rows += fmt.Sprintf(`<tr>
<td style="font-weight:600">%s</td>
<td style="font-size:12px;color:var(--dim)">%s</td>
<td>%s</td>
<td style="display:flex;gap:6px">
<form method="POST" action="/skills" style="display:inline"><input type="hidden" name="action" value="%s"><input type="hidden" name="id" value="%s"><button class="btn btn-sm" type="submit">%s</button></form>
<form method="POST" action="/skills" style="display:inline"><input type="hidden" name="action" value="uninstall"><input type="hidden" name="id" value="%s"><button class="btn btn-sm btn-d" type="submit">Uninstall</button></form>
</td></tr>`,
				templates.Esc(name), templates.Esc(desc), templates.Badge(status), toggleAction, templates.Esc(id), toggleLabel, templates.Esc(id))
		}
		installedHTML = `<table><thead><tr><th>Name</th><th>Description</th><th>Status</th><th>Actions</th></tr></thead><tbody>` + rows + `</tbody></table>`
	} else {
		installedHTML = `<div class="empty"><div class="empty-i">📦</div>No community skills installed</div>`
	}

	content := fmt.Sprintf(`<h2 class="t">Skills</h2><p class="desc">Manage builtin and community skills for your agents</p>
<div class="card" style="margin-bottom:16px"><div class="ct">Builtin Skills</div>%s</div>
<div class="card"><div class="ct">Installed Community Skills</div>%s</div>`, builtinHTML, installedHTML)

	w.Header().Set("Content-Type", "text/html")
	fmt.Fprint(w, templates.Layout("skills", s.User, content))
}

File: dashboards/express/routes/skills.js
/**
 * AgenticMail Enterprise Dashboard — Skills Routes
 * GET /skills, POST /skills/:id/enable, POST /skills/:id/disable, POST /skills/:id/uninstall
 */

const { Router } = require('express');
const { requireAuth } = require('../middleware/auth');
const { apiGet, apiPut, apiDelete } = require('../utils/api');
const { layout } = require('../views/layout');
const { buildTable } = require('../views/components/table');
const { esc, badge, statusBadge } = require('../utils/helpers');

const router = Router();

router.get('/skills', requireAuth, async (req, res) => {
  const flash = req.session.flash;
  delete req.session.flash;

  const [builtinRes, installedRes] = await Promise.all([
    apiGet('/api/engine/skills/by-category', req.session.token),
    apiGet('/api/engine/community/installed?orgId=default', req.session.token),
  ]);

  // Parse builtin skills (grouped by category)
  const categories = builtinRes.status === 200
    ? (builtinRes.body.categories || builtinRes.body || {})
    : {};

  // Parse installed community skills
  const installed = installedRes.status === 200
    ? (Array.isArray(installedRes.body.skills) ? installedRes.body.skills : (Array.isArray(installedRes.body) ? installedRes.body : []))
    : [];

  // Build builtin skills grid
  let builtinHtml = '';
  const categoryEntries = typeof categories === 'object' && !Array.isArray(categories)
    ? Object.entries(categories)
    : [];

  if (categoryEntries.length === 0) {
    builtinHtml = '<div class="empty"><span class="icon">&#9889;</span>No builtin skills found.</div>';
  } else {
    builtinHtml = categoryEntries.map(([catName, skills]) => {
      const skillsList = Array.isArray(skills) ? skills : [];
      const skillCards = skillsList.map(s => `
        <div class="skill-card" style="border:1px solid var(--border);border-radius:8px;padding:14px;background:var(--card-bg, #fff)">
          <div style="font-weight:600;font-size:14px;margin-bottom:4px">${esc(s.name || s.id || '-')}</div>
          <div style="font-size:12px;color:var(--text-muted);margin-bottom:8px;line-height:1.4">${esc(s.description || '')}</div>
          ${s.version ? `<span style="font-size:11px;color:var(--text-muted)">v${esc(s.version)}</span>` : ''}
        </div>`
      ).join('');

      return `
        <div style="margin-bottom:20px">
          <h4 style="font-size:13px;font-weight:600;color:var(--text-muted);text-transform:uppercase;letter-spacing:0.05em;margin-bottom:10px">${esc(catName)}</h4>
          <div style="display:grid;grid-template-columns:repeat(auto-fill, minmax(240px, 1fr));gap:12px">
            ${skillCards}
          </div>
        </div>`;
    }).join('');
  }

  // Build installed community skills table
  const installedRows = installed.map(s => {
    const isEnabled = s.enabled !== false && s.status !== 'disabled';
    const toggleBtn = isEnabled
      ? `<form method="post" action="/skills/${esc(s.id)}/disable" style="display:inline">
           <button class="btn btn-sm" type="submit">Disable</button>
         </form>`
      : `<form method="post" action="/skills/${esc(s.id)}/enable" style="display:inline">
           <button class="btn btn-sm btn-primary" type="submit">Enable</button>
         </form>`;
    const uninstallBtn = `<form method="post" action="/skills/${esc(s.id)}/uninstall" style="display:inline" onsubmit="return confirm('Uninstall this skill?')">
           <button class="btn btn-sm btn-danger" type="submit">Uninstall</button>
         </form>`;
    return [
      `<strong>${esc(s.name || s.id || '-')}</strong>`,
      `<span style="color:var(--text-muted)">${esc(s.version || '-')}</span>`,
      isEnabled ? badge('enabled', 'success') : badge('disabled', 'danger'),
      `${toggleBtn} ${uninstallBtn}`,
    ];
  });

  const installedTable = buildTable(
    ['Name', 'Version', 'Status', 'Actions'],
    installedRows,
    '&#128230;',
    'No community skills installed.'
  );

  const content = `
    <div class="page-header">
      <h1>Skills</h1>
      <p>Builtin capabilities and community skill extensions</p>
    </div>
    <div class="card">
      <h3>Builtin Skills</h3>
      ${builtinHtml}
    </div>
    <div class="card">
      <h3>Installed Community Skills (${installed.length})</h3>
      ${installedTable}
    </div>`;

  res.send(layout('skills', req.session.user, content, flash));
});

router.post('/skills/:id/enable', requireAuth, async (req, res) => {
  const result = await apiPut(`/api/engine/community/skills/${req.params.id}/enable`, req.session.token, {});

  if (result.status < 300) {
    req.session.flash = { message: 'Skill enabled', type: 'success' };
  } else {
    req.session.flash = { message: (result.body && result.body.error) || 'Failed to enable skill', type: 'danger' };
  }
  res.redirect('/skills');
});

router.post('/skills/:id/disable', requireAuth, async (req, res) => {
  const result = await apiPut(`/api/engine/community/skills/${req.params.id}/disable`, req.session.token, {});

  if (result.status < 300) {
    req.session.flash = { message: 'Skill disabled', type: 'success' };
  } else {
    req.session.flash = { message: (result.body && result.body.error) || 'Failed to disable skill', type: 'danger' };
  }
  res.redirect('/skills');
});

router.post('/skills/:id/uninstall', requireAuth, async (req, res) => {
  const result = await apiDelete(`/api/engine/community/skills/${req.params.id}/uninstall`, req.session.token);

  if (result.status < 300) {
    req.session.flash = { message: 'Skill uninstalled', type: 'success' };
  } else {
    req.session.flash = { message: (result.body && result.body.error) || 'Failed to uninstall skill', type: 'danger' };
  }
  res.redirect('/skills');
});

module.exports = router;
